kernel/src/util/iovec.rs

```diff
@@ -37,92 +39,190 @@ impl TryFrom<UserIoVec> for IoVec {
+pub struct IoVecReader<'a>(Box<[VmReader<'a>]>);
 
-        CurrentUserSpace::get().read_bytes(self.base, &mut VmWriter::from(dst))
+/// A collection of [`VmWriter`]s.
+///
+/// Such writers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecWriter<'a>(Box<[VmWriter<'a>]>);
+
+impl<'a> IoVecReader<'a> {
+    /// Creates a new `IoVecReader` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let readers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::reader)?;
+        Ok(Self(readers))
     }
 
-    /// Writes bytes from the `src` buffer
-    /// to the user space buffer pointed by the `IoVec`.
-    ///
-    /// If successful, the written length will be equal to `src.len()`.
-    ///
-    /// # Panics
-    ///
-    /// This method will panic if
-    /// 1. `src.len()` is not the same as `self.len()`;
-    /// 2. `self.is_empty()` is `true`.
-    pub fn write_exact_to_user(&self, src: &[u8]) -> Result<()> {
-        assert_eq!(src.len(), self.len);
-        assert!(!self.is_empty());
+    pub fn readers_mut(&'a mut self) -> &'a mut [VmReader<'a>] {
+        &mut self.0
+    }
+}
 
-        CurrentUserSpace::get().write_bytes(self.base, &mut VmReader::from(src))
+impl<'a> IoVecWriter<'a> {
+    /// Creates a new `IoVecWriter` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let writers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::writer)?;
+        Ok(Self(writers))
     }
 
-    /// Reads bytes to the `dst` buffer
-    /// from the user space buffer pointed by the `IoVec`.
-    ///
-    /// If successful, returns the length of actually read bytes.
-    pub fn read_from_user(&self, dst: &mut [u8]) -> Result<usize> {
-        let len = self.len.min(dst.len());
-        CurrentUserSpace::get().read_bytes(self.base, &mut VmWriter::from(&mut dst[..len]))?;
-        Ok(len)
+    pub fn writers_mut(&'a mut self) -> &'a mut [VmWriter<'a>] {
+        &mut self.0
     }
+}
 
-    /// Writes bytes from the `src` buffer
-    /// to the user space buffer pointed by the `IoVec`.
+/// Trait defining the read behavior for a collection of [`VmReader`]s.
+pub trait IoVecRead {
+    /// Reads data to the provided writer from all collected [`VmReader`]s,
+    /// accumulating total bytes read.
     ///
-    /// If successful, returns the length of actually written bytes.
-    pub fn write_to_user(&self, src: &[u8]) -> Result<usize> {
-        let len = self.len.min(src.len());
-        CurrentUserSpace::get().write_bytes(self.base, &mut VmReader::from(&src[..len]))?;
-        Ok(len)
+    /// This method returns [`Errno::EFAULT`]
+    /// if a page fault occurs and _not all bytes_ are successfully read.
+    fn read_exact(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;
```

Comment:
I don't know why this method should be called `read_exact`. The word "exact" does not even appear in the Rust doc.

```rust
/// Read the exact number of bytes required to fill `buf`.
fn read_exact(&mut self, buf: &mut [u8]) -> Result<()> {
```

You see: in Rust std, the semantics described by the Rust doc matches with the method name. But this is not the case for our `read_exact`.

So what is the semantics of our so-called `read_exact`? 

```rust
impl<'a> IoVecRead for VmReader<'a> {
    fn read_exact(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize> {
        match self.read_fallible(writer) {
            Ok(copied_len) => Ok(copied_len),
            Err(_) => return_errno_with_message!(
                Errno::EFAULT,
                "cannot copy from `VmReader` to `VmWriter`"
            ),
        }
    }
}
```

According to the implementation above, it is the same as `read_fallible`...

The `write_all` method has the same problem. Its semantics is not clear from the method name and the implementation.

Reply:
I have renamed the current method as `read_to` and `write_from`. 

@lrh2000, thanks for your explanation! 

The current API reads until either the writer is filled, or the `VmReaderArray` is empy. So the read len should be the smaller length of writer and  reader array. Emmmm, I can't insert an `exact` into the documents. So I switch to the old name.

> read_exact will return Err(EFAULT) if a page fault occurs in the middle. This means the user does not want to use the partially copied buffer. 

Yes. I think this is the usual case in our kernel, so it may deserve short name.

> read_to will return Err(EFAULT) if a page fault occurs in the middle and no bytes have been copied. This means the user does want to use the partially copied buffer.

Maybe we can give it a long name like `maybe_partial_read` in the future?

Reply:
> can't insert an exact

Well, at least:
```rust
/// Reads the exact number of bytes required to exhaust `self` or fill `buf`.
```

Reply:
Like so many bad names in Linux, `read_to` is a bad one that we should not follow.

```rust
    /// Reads data to the provided writer from all collected [`VmReader`]s,
    /// accumulating total bytes read.
    ///
    /// This method returns [`Errno::EFAULT`]
    /// if a page fault occurs and _not all bytes_ are successfully read.
    fn read_to(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;
```

So according to the Rust doc of the latest version, "to" means "to the provided buffer". If this is what "to" means, then most our `read`-family methods of `VmReader` should begin their names with `read_to_` because they all read to a writer or slice.

Here is my suggestions:
* The original `read_exact` should be named `read_to_end`. On success, the method guarantees that all bytes are read from the source `VmReader`s to a target `VmWriter`. The method fails if one of the two conditions is met: (1) the target writer does not have enough room to store all the bytes; or (2) any page fault occurs while data is copied from the source to the target.
* The original `read_to` should be named `read_some`. On success, the method guarantees that some bytes are read from the source `VmReader`s to a target `VmWriter` for all non-trivial cases. The trivial case is when all source readers have remaining data (`reader.remain() == 0` and the target writer has no available room for new data (`writer.avail() == 0`). On failure, a page fault must occur and no data are copied at all.

Reply:
> the target writer does not have enough room to store all the bytes

This will occur frequently. The userspace may supply `IoVec`s whose length exceeds the length of the kernel buffer (e.g., the ring buffer of pipes). What's your suggestion to deal with this situation?

Reply:
I don't have an easy answer. It seems that there are many different possible usage scenarios. I think we need to sit back and think more thoroughly about our needs and design the API accordingly.

Reply:
I don't think we have too many choices. We can just go through them one by one:

Keep the original semantics:
```rust
/// Reads the exact number of bytes required to exhaust `self` or fill `writer`.
```
  - The downside is that it is not clear whether `self` is exhausted or `writer` is filled after the method returns. This is especially true when comparing with the following choices.
  - However:
    - I think there are enough use cases shown in this PR where we just don't care which one happens.
    - The only two use cases that care which one happen is:
      - (i) When copying to the the tailing part of the ring buffer, the want to ensure the tail part is failled (so that we can copy the remaining to the heading part later);
      - (ii) When copying UDP messages, we want to ensure the whole UDP message is copied.
      - Our current stragety is to check the length before starting copying. While I'd like to see this gets enforced in the API level, it is not clear to me how to do it.
    - The same downside also applies for `VmReader::read_faillible`.

Require the `writer` to be filled (this is equivalent to `Read::read_exact`):
```rust
/// Reads the exact number of bytes required to fill `writer`.
```
 - The downside is that we have to call `sum_lens` first and then call `writer.limit(len)`. This just seems redundant, since in most cases we don't care whether the `writer` will be filled or not.

Require the `self` to be exhausted (this is equivalent to `Read::read_to_end`):
```rust
/// Reads the exact number of bytes required to exhaust `self`.
```
 - Unless we implement `MultiRead::limit`, we cannot resolve https://github.com/asterinas/asterinas/pull/1294#discussion_r1759665365. I don't want to add the `MultiRead::limit` method because it is just redundant in most cases.

Or, we can have a combination of the above methods.

Reply:
> Keep the original semantics
> /// Reads the exact number of bytes required to exhaust `self` or fill `writer`.

I think the original semantic is just the same as [Read::read](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#tymethod.read). How about just defining the method as

```rust
pub trait MultiRead {
    /// Writes the exact number of bytes required to exhaust `writer` or fill `self`,
    /// accumulating total bytes read.
    ///
    /// If the return value is `Ok(n)`,
    /// then `n` should be `min(self.sum_lens(), reader.remain())`.
    ///
    /// # Errors
    ///
    /// This method returns [`Errno::EFAULT`] if a page fault occurs.
    fn read(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;
}
```

Reply:
> I think the original semantic is just the same as [Read::read](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#tymethod.read).

`Read::read` is allowed to return `Ok(n)` where `0 < n < min(self.sum_lens(), writer.avail())`. This API doesn't allow it.

Earlier I wanted to use the word "exact" to tell users that this is not possible. While I still think the word "exact" is the correct one to use here (at least I cannot think of out a better one?), `read` is acceptable to me if we cannot reach a consensus on this matter.